staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1291,12 +1291,12 @@ func (c *cacheWatcher) convertToWatchEvent(event *watchCacheEvent) *watch.Event
 
 	switch {
 	case curObjPasses && !oldObjPasses:
-		return &watch.Event{Type: watch.Added, Object: getEventObject(event.Object)}
+		return &watch.Event{Type: watch.Added, Object: event.Object}
 	case curObjPasses && oldObjPasses:
-		return &watch.Event{Type: watch.Modified, Object: getEventObject(event.Object)}
```

Comment:
why drop these calls? if we're already holding CachedObject instances, and we're confident nothing further on is mutating, these should be no-op checks, right?

Reply:
It's not always CachingObject - we're using only for non-negligible amount of watchers:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L856

Reply:
so... aren't there still potentially mutations of these objects we need to guard against? I'm thinking especially of https://github.com/kubernetes/kubernetes/issues/82497 (where we mutate objects briefly when encoding them)

Reply:
Hmm - interesting. I don't fully understand why we're not seeing this as an issue in presubmits (it seems like something that we should hit in performance tests presubmit), but ignoring that, I agree it actually seems like it could cause issues.

So I think, the safer step (at least as a first one) would be to:
- given that that cachingObject is now not doing any deep-copy at the beginning (just doing that lazily if really needed, which according to my knowledge should never happen) switch to using cachingObject unconditionally in all places:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L856

- and with that it's safe, because the actual encoding that is happening on cachingObject is actually performing a deepcopy underneath:
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go#L127

Does that sound better to you?

Reply:
In the meantime I added a third commit that implements this - PTAL